### PR TITLE
Cross-platform build.sh

### PR DESCRIPTION
**Pull request by str4d**
This enables `zcutil/build.sh` to be used on MacOS, and likely other non-Linux platforms.

### CODE REVIEW

**Comment by str4d** [comment]
@zkbot try

**Comment by str4d** [comment]
Currently this just fixes the `readlink` issue in `build.sh`. We also need to fix the issue of `HOST` and `BUILD` defaulting to Linux strings. `configure.ac` internally does host detection, and the depends system similarly uses `config.guess`, so I'm wondering whether we even need to specify defaults for `HOST` and `BUILD` at all?

**Comment by radix42** [comment]
the depends system cannot detect the correct values for HOST and BUILD for cross-compilation, for instance. non-linux platforms are all passing both values into the depends system via env vars

**Comment by str4d** [comment]
Okay, so for cross-compiling we will need to have a separate test path in the CI server. But anything that is not cross-compiled, it should be possible to not need to set env vars, at least for supported platform.

Force-pushed to tweak the first commit, and added another commit that uses the depends system's `config.guess` instead of hard-coding `HOST` and `BUILD` defaults.

**Comment by str4d** [comment]
Before merging this, we should check that builds are still deterministic. This *should* be the case, as long as `config.guess` is deterministic.

**Comment by bitcartel** [comment]
I am going to check to see if builds are deterministic or not.

**Comment by bitcartel** [comment]
In Gitian, I get an error trying to build.

```
make[3]: Entering directory '/home/debian/build/zcash/src/snark'
g++ -o src/algebra/curves/alt_bn128/alt_bn128_g1.o   src/algebra/curves/alt_bn128/alt_bn128_g1.cpp -c -MMD -fPIC -DBINA
RY_OUTPUT -DNO_PT_COMPRESSION=1 -std=c++11 -Wall -Wextra -Wno-unused-parameter -Wno-comment -Wfatal-errors -O2 -march=x
86-64 -DMONTGOMERY_OUTPUT -DCURVE_ALT_BN128 -I//include -Isrc -DNO_PROCPS -static -DSTATIC -DMULTICORE -fopenmp
In file included from src/algebra/fields/fp.hpp:13:0,
                 from src/algebra/curves/alt_bn128/alt_bn128_init.hpp:11,
                 from src/algebra/curves/alt_bn128/alt_bn128_g1.hpp:11,
                 from src/algebra/curves/alt_bn128/alt_bn128_g1.cpp:8:
src/algebra/fields/bigint.hpp:14:17: fatal error: gmp.h: No such file or directory
 #include <gmp.h>
                 ^
compilation terminated.
```
I tagged this branch in my PR and set gitian to build the tag `macos-tag` : https://github.com/bitcartel/zcash/tree/macos-tag

**Comment by str4d** [comment]
Tried the parent commit of this PR, and it fails too. I'll rebase and test again.

**Comment by str4d** [comment]
Rebasing fixes the problem. Also, I realise that my original question of determinism was completely wrong - the Gitian build doesn't use `zcutil/build.sh`, and thus is completely unaffected by this PR.

File: zcutil/build.sh [hunk 1]
```diff
@@ -2,19 +2,35 @@
...
-#   MAKE=gmake ./zcutil/build.sh -j$(nproc)
+#   MAKE=custom-make ./zcutil/build.sh -j$(nproc)
 if [[ -z "${MAKE-}" ]]; then
-    MAKE=make
+    gprefix MAKE make
 fi
 
 # Allow overrides to $BUILD and $HOST for porters. Most users will not need it.
 #   BUILD=i686-pc-linux-gnu ./zcutil/build.sh
 if [[ -z "${BUILD-}" ]]; then
```
Comment (arielgabizon) [inline zcutil/build.sh]: what does this syntax mean in bash?

Comment (str4d) [reply zcutil/build.sh]: It causes the brackets to expand to the empty string if `BUILD` is unset. See https://stackoverflow.com/a/32964575

Comment (arielgabizon) [reply zcutil/build.sh]: and empty string is considered true for `if`?

**Comment by bitcartel** [comment]
Same problem, using latest branch.
```
src/algebra/fields/bigint.hpp:14:17: fatal error: gmp.h: No such file or directory
 #include <gmp.h>
                 ^
compilation terminated.
```

File: zcutil/build.sh (hunk 1 above)
Comment (arielgabizon) [reply zcutil/build.sh]: ok the `-z` turns an empty string to true.

**Comment by arielgabizon** [comment]
I don't understand what the content has to do with the PR name.
Main things here are use `gmake` instead of `make` if possible, which is also supported by linux.
And using `config.guess` instead of assuming it's  `linux-gnu` which is also not specific to mac os

**Comment by str4d** [comment]
@bitcartel could you please confirm what commit you rebuilt with? I was able to recreate your Gitian issue, and successfully resolved it after rebasing this PR.

@arielgabizon the relation is, these changes are needed for `zcutil/build.sh` to be usable on both Linux and MacOS without per-system environment variable modifications (which is my goal, to make building simpler for both users and the CI system). So e.g. using `config.guess` is not specific to MacOS, but is necessary because the previous hard-coded string *would not work* on MacOS. The intent of this PR is that it enables MacOS support, without degrading the existing Linux support (and this will be checked by the CI system).

**Comment by arielgabizon** [comment]
@str4d is preferring `gmake` related to mac os support?

**Comment by arielgabizon** [comment]
So, I'm wondering if a better name for the PR is "Extend support for build.sh beyond linux-gnu"?

**Comment by str4d** [comment]
It was, yes - on MacOS, `make`, `gcc` etc. point to non-GNU tools, and if you install the GNU ones they are prefixed with a g. However, you raise a good point - I need to check whether this is still necessary, as I changed strategy in #2674 to instead use their default tools. If it isn't necessary, I will remove it from this PR.

**Comment by str4d** [comment]
Okay, I recall what I did. I needed this g-prefix change to fix a problem with readlink, which behaves completely differently on MacOS, so you need greadlink. I decided to migrate the make command at the same time, as some people use gmake. However, it isn't strictly necessary, and I agree it is confusing, so I will back out that part of the change.

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Works on CI!

- [Before](https://ci.z.cash/#/builders/16/builds/56) - fails at readlink
- [After](https://ci.z.cash/#/builders/16/builds/57) - fails in dependencies

**Comment by str4d** [comment]
Manually cancelled early because one of the required builders failed with an unrelated transient AWS failure.

@zkbot retry

**Pull request merged**